controllers/cloud.redhat.com/clowdapp_controller.go

```diff
@@ -138,192 +133,24 @@ func (r *ClowdAppReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
-	}
-
-	if ens.ObjectMeta.DeletionTimestamp != nil {
-		log.Info("Env target namespace is to be deleted - skipping reconcile")
-		return ctrl.Result{}, nil
-	}
-
-	ans := &core.Namespace{}
-	if getAppNSErr := r.Client.Get(ctx, types.NamespacedName{Name: app.Namespace}, ans); getAppNSErr != nil {
-		return ctrl.Result{Requeue: true}, getAppNSErr
-	}
-
-	if ans.ObjectMeta.DeletionTimestamp != nil {
-		log.Info("App namespace is to be deleted - skipping reconcile")
-		return ctrl.Result{}, nil
-	}
-
-	if env.Generation != env.Status.Generation {
-		r.Recorder.Eventf(&app, "Warning", "ClowdEnvNotReconciled", "Clowder Environment [%s] is not reconciled", app.Spec.EnvName)
-		log.Info("Env not yet reconciled")
-		if setClowdStatusErr := SetClowdAppConditions(ctx, r.Client, &app, crd.ReconciliationFailed, fmt.Errorf("clowd env not reconciled")); setClowdStatusErr != nil {
-			log.Info("Set status error", "err", setClowdStatusErr)
-			return ctrl.Result{Requeue: true}, setClowdStatusErr
-		}
-		return ctrl.Result{}, nil
-	}
-
-	if !env.IsReady() {
-		r.Recorder.Eventf(&app, "Warning", "ClowdEnvNotReady", "Clowder Environment [%s] is not ready", app.Spec.EnvName)
-		log.Info("Env not yet ready")
-		if setClowdStatusErr := SetClowdAppConditions(ctx, r.Client, &app, crd.ReconciliationFailed, fmt.Errorf("clowd env not ready")); setClowdStatusErr != nil {
-			log.Info("Set status error", "err", setClowdStatusErr)
-			return ctrl.Result{Requeue: true}, setClowdStatusErr
-		}
-
-		return ctrl.Result{Requeue: true}, nil
-	}
-
 	cacheConfig := rc.NewCacheConfig(Scheme, errors.ClowdKey("log"), ProtectedGVKs, DebugOptions)
-
 	cache := rc.NewObjectCache(ctx, r.Client, cacheConfig)
-	//cache := providers.NewObjectCache(ctx, r.Client, scheme)
-
-	provider := providers.Provider{
-		Client: r.Client,
-		Ctx:    ctx,
-		Env:    &env,
-		Cache:  &cache,
-		Log:    log,
-	}
 
-	if provErr := r.runProviders(log, &provider, &app); provErr != nil {
-		r.Recorder.Eventf(&app, "Warning", "FailedReconciliation", "Clowdapp requeued [%s]", app.GetClowdName())
-		if setClowdStatusErr := SetClowdAppConditions(ctx, r.Client, &app, crd.ReconciliationFailed, provErr); setClowdStatusErr != nil {
-			log.Info("Set status error", "err", setClowdStatusErr)
-			return ctrl.Result{Requeue: true}, setClowdStatusErr
-		}
-		log.Info("Provider error", "err", provErr)
-		return ctrl.Result{Requeue: true}, provErr
-	}
-
-	cacheErr := cache.ApplyAll()
+	log.Info("Reconciliation started")
 
-	if cacheErr != nil {
-		r.Recorder.Eventf(&app, "Warning", "FailedReconciliation", "Clowdapp requeued [%s]", app.GetClowdName())
-		if setClowdStatusErr := SetClowdAppConditions(ctx, r.Client, &app, crd.ReconciliationFailed, cacheErr); setClowdStatusErr != nil {
-			log.Info("Set status error", "err", setClowdStatusErr)
-			return ctrl.Result{Requeue: true}, setClowdStatusErr
+	reconciliation := MakeClowdAppReconciliation(ctx, &cache, r.Client, r.Recorder, &app, log, &req)
```

Comment:
Yeah I tried to avoid judgement calls where possible but this was one of them so I figured I'd try it and see what happens.

I don't have any kind of mental model for resource usage and performance impacts re: Clowder running in real life. Other than "try to use as few resources as possible" I don't have any sort of heuristics around that in my head about like what is and is not expensive relatively speaking. So in moving this code I didn't know if I was doing something as cheap as allocating an empty struct or as expensive as spinning up some complex system. Considering the comment I suspect its more like the latter than the former.

I'll change this to be a step method in `ClowdAppReconciliation` and I'll do it as late as possible in the process.

Reply:
Moved it to a step!